sensors/server-include/libsensors/sensor.h

```diff
@@ -18,18 +18,68 @@
 
 #include <sys/rb.h>
 #include <posix/idtree.h>
+#include <spi-msg.h>
 
 
 #define THREAD_PRIORITY_MSGSRV 3 /* priority of messaging server of sensorhub */
 #define THREAD_PRIORITY_SENSOR 3 /* priority of sensor driver threads */
 
+enum sensor_bus_type { bus_none = 0,
+	bus_spi,
+	bus_i2c };
+
+typedef struct _sensor_bus_t {
+	enum sensor_bus_type bustype; /* bus type identifier for the device for correct initialisation procedure */
+	struct {
+		/**
+		 * bus transaction
+		 *
+		 * out - output data buffer
+		 * olen - length of output data buffer
+		 * in - input data buffer
+		 * ilen - input data buffer length
+		 * iskip - data skip size during read
+		 *
+		 * Returns 0 on success, -1 on failure.
+		 */
+		int (*bus_xfer)(const struct _sensor_bus_t *bus, const void *out, size_t olen, void *in, size_t ilen, size_t iskip);
+
+		/**
+		 * bus configuration function.
+		 *
+		 * For SPI this sets the clock speed in Hz and SPI mode character.
+		 * For I2C this only sets speed in Hz, mode is unused.
+		 *
+		 * Is NULL when not supported.
+		 * Returns 0 on success, -1 on bus donfiguration failure.
+		 */
+		int (*bus_cfg)(struct _sensor_bus_t *bus, int speed, char mode);
+		int (*bus_close)(struct _sensor_bus_t *bus);
+
+		/**
+		 * bus open function accepts dev/subdev parameters.
+		 *
+		 * For SPI these are the name, or virtual path to spi device (e.g "/dev/spi1"), and subdev is the `ss` number.
+		 * For I2C these are the name, or virtual path to i2c device (e.g "/dev/i2c1"), and subdev is unused.
+		 */
+		int (*bus_open)(struct _sensor_bus_t *bus, const char *dev, const char *subdev);
+	} ops;
+	union {
+		struct {
+			spimsg_ctx_t ctx; /* spi context, not to be used by device */
+			oid_t ss;         /* spi ss oid, not to be used by device */
+		} spi;
+	} spec;
```

Comment:
We can hide details of the bus context by using a pointer and forward declaration instead. Then thanks to that each bus driver might use its specific context implementation.

```
struct bus_ctx;
(...)
struct bus_ctx *ctx;
```

or (former preferred imho)

```
struct bus_ctx;
typedef struct bus_ctx bus_ctx_t;
(...)
bus_ctx_t *ctx;
```

Implementation is then provided by the lib user and bus provider.

Reply:
Implemented in the first form proposed